x-pack/plugins/observability_solution/observability_ai_assistant/server/service/client/adapters/openai_adapter.ts

```diff
@@ -123,7 +124,10 @@ export const createOpenAiAdapter: LlmApiAdapterFactory = ({
           ...(!!functionsForOpenAI?.length
             ? {
                 tools: functionsForOpenAI.map((fn) => ({
-                  function: pick(fn, 'name', 'description', 'parameters'),
+                  function:
```

Comment:
I think we can go for a different fix that doesn't special-case the get_data_on_screen function. Instead of using the function description, we can add an instruction in the same place where we are registering the get_data_on_screen function: https://github.com/elastic/kibana/blob/ee341d5f801ca42ed26acf0544b0bc59948d0214/x-pack/plugins/observability_solution/observability_ai_assistant/server/service/chat_function_client/index.ts#L51.

Reply:
Sounds good. Thanks @dgieselaar 
I'm looking into it.

Reply:
Hi @dgieselaar 

I implemented the above change as follows:
```
if (allData.length) {
      this.registerFunction(
        {
          name: GET_DATA_ON_SCREEN_FUNCTION_NAME,
          description: 'Get data that is on the screen',
          visibility: FunctionVisibility.AssistantOnly,
          parameters: {
            type: 'object',
            properties: {
              data: {
                type: 'array',
                description:
                  'The pieces of data you want to look at it. You can request one, or multiple',
                items: {
                  type: 'string',
                  enum: allData.map((data) => data.name),
                },
              },
            },
            required: ['data' as const],
          },
        },
        async ({ arguments: { data: dataNames } }) => {
          return {
            content: allData.filter((data) => dataNames.includes(data.name)),
          };
        },
        ['all']
      );

      const instruction: RegisterInstructionCallback = ({ availableFunctionNames }) =>
        availableFunctionNames.includes(GET_DATA_ON_SCREEN_FUNCTION_NAME)
          ? `The ${GET_DATA_ON_SCREEN_FUNCTION_NAME} function will ${dedent(`Get data that is on the screen:
            ${allData.map((data) => `${data.name}: ${data.description}`).join('\n')}
          `)}`
          : undefined;

      this.registerInstruction({ instruction, scopes: ['all'] });
    }
```

However, when I register the instruction when the function is being registered, the instruction is appended to the top of the prompt, as follows:

```
{
      role: 'system',
      content: 'The get_data_on_screen function will Get data that is on the screen:\n' +
        '.es-query: An available rule is Elasticsearch query.\n' +
        'observability.rules.custom_threshold: An available rule is Custom threshold.\n' +
        'xpack.ml.anomaly_detection_alert: An available rule is Anomaly detection.\n' +
        'slo.rules.burnRate: An available rule is SLO burn rate.\n' +
        'metrics.alert.threshold: An available rule is Metric threshold.\n' +
        'metrics.alert.inventory.threshold: An available rule is Inventory.\n' +
        'logs.alert.document.count: An available rule is Log threshold.\n' +
        'xpack.uptime.alerts.tlsCertificate: An available rule is Uptime TLS.\n' +
        'xpack.uptime.alerts.monitorStatus: An available rule is Uptime monitor status.\n' +
        'xpack.uptime.alerts.durationAnomaly: An available rule is Uptime Duration Anomaly.\n' +
        'xpack.synthetics.alerts.monitorStatus: An available rule is Synthetics monitor status.\n' +
        'xpack.synthetics.alerts.tls: An available rule is Synthetics TLS certificate.\n' +
        'apm.error_rate: An available rule is Error count threshold.\n' +
        'apm.transaction_error_rate: An available rule is Failed transaction rate threshold.\n' +
        'apm.transaction_duration: An available rule is Latency threshold.\n' +
        'apm.anomaly: An available rule is APM Anomaly.\n' +
        '\n' +
        'You are a helpful assistant for Elastic Observability. Your goal is to help the Elastic Observability users to quickly assess what is happening in their observed systems. You can help them visualise and analyze data, investigate their systems, perform root cause analysis or identify optimisation opportunities.\n' +
        '\n' +
        "  It's very important to not assume what the user is meaning. Ask them for clarification if needed.\n" +
        '\n' +
        '  If you are unsure about which function should be used and with what arguments, ask the user for clarification or confirmation.\n' +
        '\n' +
        `  In KQL ("kqlFilter")) escaping happens with double quotes, not single quotes. Some characters that need escaping are: ':()\\  /". Always put a field value in double quotes. Best: service.name:"opbeans-go". Wrong: service.name:opbeans-go. This is very important!\n` +
        '\n' +
        '  You can use Github-flavored Markdown in your responses. If a function returns an array, consider using a Markdown table to format the response.\n' +
        '\n' +
        '  Note that ES|QL (the Elasticsearch Query Language which is a new piped language) is the preferred query language.\n' +
        '\n' +
        '  If you want to call a function or tool, only call it a single time per message. Wait until the function has been executed and its results\n' +
        '  returned to you, before executing the same tool or another tool again if needed.\n' +
        '\n' +
        '  DO NOT UNDER ANY CIRCUMSTANCES USE ES|QL syntax (`service.name == "foo"`) with "kqlFilter" (`service.name:"foo"`).\n' +
        '\n' +
        '  The user is able to change the language which they want you to reply in on the settings page of the AI Assistant for Observability, which can be found in the Stack Management app under the option AI Assistants.\n' +
        '  If the user asks how to change the language, reply in the same language the user asked in.\n' +
        '\n' +
        'You MUST use the "query" function when the user wants to:\n' +
        '  - visualize data\n' +
        '  - run any arbitrary query\n' +
        '  - breakdown or filter ES|QL queries that are displayed on the current page\n' +
        '  - convert queries from another language to ES|QL\n' +
        '  - asks general questions about ES|QL\n' +
        '\n' +
        '  DO NOT UNDER ANY CIRCUMSTANCES generate ES|QL queries or explain anything about the ES|QL query language yourself.\n' +
        '  DO NOT UNDER ANY CIRCUMSTANCES try to correct an ES|QL query yourself - always use the "query" function for this.\n' +
        '\n' +
        '  If the user asks for a query, and one of the dataset info functions was called and returned no results, you should still call the query function to generate an example query.\n' +
        '\n' +
        '  Even if the "query" function was used before that, follow it up with the "query" function. If a query fails, do not attempt to correct it yourself. Again you should call the "query" function,\n' +
        '  even if it has been called before.\n' +
        '\n' +
        '  When the "visualize_query" function has been called, a visualization has been displayed to the user. DO NOT UNDER ANY CIRCUMSTANCES follow up a "visualize_query" function call with your own visualization attempt.\n' +
        '  If the "execute_query" function has been called, summarize these results for the user. The user does not see a visualization in this case.\n' +
        '\n' +
        'You MUST use the "get_dataset_info"  function before calling the "query" or the "changes" functions.\n' +
        '\n' +
        'If a function requires an index, you MUST use the results from the dataset info functions.\n' +
        '\n' +
        'You have access to data on the screen by calling the "get_data_on_screen" function.\n' +
        'Use it to help the user understand what they are looking at. A short summary of what they are looking at is available in the return of the "context" function.\n' +
        'Data that is compact enough automatically gets included in the response for the "context" function.\n' +
        '\n' +
        'You do not have a working memory. If the user expects you to remember the previous conversations, tell them they can set up the knowledge base.',
      function_call: undefined,
      name: undefined
    }
```

Reply:
right, @sphilipse you also have a requirement to treat certain instructions differently, correct? is that about ordering or some other mechanism?

Reply:
Nah that was about the way we use scopes to filter out some things in some cases, but I think that's resolved with the work we're doing in https://github.com/elastic/kibana/pull/196322.

We already have ways to fix ordering, right? System instructions are added to the top, but there are separate instructions that are added lower. You just need to not register them as an AdHocInstruction (see: https://github.com/elastic/kibana/blob/main/x-pack/plugins/observability_solution/observability_ai_assistant/common/types.ts and https://github.com/elastic/kibana/blob/main/x-pack/plugins/observability_solution/observability_ai_assistant/server/service/util/get_system_message_from_instructions.ts)

Reply:
Thank you @sphilipse 
I've updated the PR with the implementation to register adhoc instructions.

With this implementation, I am able to register the description of the `get_data_on_screen` function as an adhoc instruction and order it at the end of the system message as specified by the `getSystemMessageFromInstructions` function.

cc: @dgieselaar